src/deluge/gui/views/session_view.cpp

```diff
@@ -2527,3 +2823,889 @@ void SessionView::modEncoderAction(int32_t whichModEncoder, int32_t offset) {
+				clipPressEnded();
+			}
+
+			// Second finger up, clone clip
+			else if (gridSecondPressedX == x && gridSecondPressedY == y) {
+				gridClonePad(gridFirstPressedX, gridFirstPressedY, gridSecondPressedX, gridSecondPressedY);
+				gridResetPresses(); // Also reset first press so clip does not get armed
+				gridPreventArm = false;
+				clipPressEnded();
+			}
+		}
+	}
+
+	if (currentUIMode != UI_MODE_EXPLODE_ANIMATION) {
+		requestRendering(this, 0xFFFFFFFF, 0xFFFFFFFF);
+		view.flashPlayEnable();
+	}
+	return ActionResult::DEALT_WITH;
+}
+
+ActionResult SessionView::gridHandleScroll(int32_t offsetX, int32_t offsetY) {
+	gridResetPresses();
+
+	// Fix the range
+	currentSong->songGridScrollY =
+	    std::clamp<int32_t>(currentSong->songGridScrollY + offsetY, 0, kMaxNumSections - kGridHeight);
+	currentSong->songGridScrollX = std::clamp<int32_t>(currentSong->songGridScrollX + offsetX, 0,
+	                                                   std::max<int32_t>(0, (gridTrackCount() - kDisplayWidth) + 1));
+
+	// This is the right place to add new features like moving clips or tracks :)
+
+	requestRendering(this, 0xFFFFFFFF, 0xFFFFFFFF);
+	view.flashPlayEnable();
+	return ActionResult::DEALT_WITH;
+}
+
+void SessionView::gridTransitionToSessionView() {
+	Sample* sample;
+
+	if (currentSong->currentClip->type == CLIP_TYPE_AUDIO) {
+		// If no sample, just skip directly there
+		if (!((AudioClip*)currentSong->currentClip)->sampleHolder.audioFile) {
+			changeRootUI(&sessionView);
+			memcpy(PadLEDs::imageStore, PadLEDs::image, sizeof(PadLEDs::image));
+			finishedTransitioningHere();
+			return;
+		}
+	}
+
+	currentUIMode = UI_MODE_EXPLODE_ANIMATION;
+
+	memcpy(PadLEDs::imageStore[1], PadLEDs::image, (kDisplayWidth + kSideBarWidth) * kDisplayHeight * 3);
+	memcpy(PadLEDs::occupancyMaskStore[1], PadLEDs::occupancyMask, (kDisplayWidth + kSideBarWidth) * kDisplayHeight);
+	if (getCurrentUI() == &instrumentClipView) {
+		instrumentClipView.fillOffScreenImageStores();
+	}
+
+	auto clipX = std::clamp<int32_t>(
+	    gridXFromTrack(gridTrackIndexFromTrack(currentSong->currentClip->output, gridTrackCount())), 0, kDisplayWidth);
+	auto clipY = std::clamp<int32_t>(gridYFromSection(currentSong->currentClip->section), 0, kDisplayHeight);
+
+	if (currentSong->currentClip->type == CLIP_TYPE_AUDIO) {
+		waveformRenderer.collapseAnimationToWhichRow = clipY;
+
+		PadLEDs::setupAudioClipCollapseOrExplodeAnimation((AudioClip*)currentSong->currentClip);
+	}
+	else {
+		PadLEDs::explodeAnimationYOriginBig = clipY << 16;
+	}
+
+	PadLEDs::explodeAnimationXStartBig = clipX << 16;
+	PadLEDs::explodeAnimationXWidthBig = (1 << 16);
+
+	PadLEDs::recordTransitionBegin(kClipCollapseSpeed);
+	PadLEDs::explodeAnimationDirection = -1;
+
+	if (getCurrentUI() == &instrumentClipView) {
+		PadLEDs::clearSideBar();
+	}
+
+	PadLEDs::explodeAnimationTargetUI = this;
+	uiTimerManager.setTimer(TIMER_MATRIX_DRIVER, 35);
+}
+
+void SessionView::gridTransitionToViewForClip(Clip* clip) {
```

Comment:
If clip can't be null, should use a reference here.

Reply:
Fully agree with using references wherever something can not be null however in this case we supply a pointer and mostly cast it to something else which makes it look very weird. Gonna resolve this as discussed in Discord